src/CodeGen_LLVM.cpp

```diff
@@ -4136,6 +4288,22 @@ void CodeGen_LLVM::visit(const Shuffle *op) {
     }
 }
 
+void CodeGen_LLVM::visit(const Atomic *op) {
+    if (op->mutex_name != "") {
+        internal_assert(op->mutex_indices.size() <= 1) << "Atomic mutex access index should be flattened.";
+        Expr index = op->mutex_indices.size() == 1 ? op->mutex_indices[0] : Expr(0);
+        user_assert(index.type().is_scalar()) << "The vectorized atomic operation in " <<
+            op->producer_name << " is lowered into a mutex lock, which does not support vectorization.\n";
+
+        // Mutex locking & unlocking are handled by function calls generated by previous lowering passes.
+        codegen(op->body);
+    } else {
+        emit_atomic_stores = true;
```

Comment:
I think this will break with nested atomic nodes (if that's even allowed). Use:
```
{
ScopedValue<bool> old_emit_atomic_stores(emit_atomic_stores, true);
codegen(op->body);
}
```

That will restore the old value of emit_atomic_stores instead of setting it to false.

Alternatively, assert !emit_atomic_stores before setting it to true if nested atomic nodes are not allowed.

Reply:
Nested atomics without mutex locks is fine. Nested mutex locks lead to deadlocks. I added checks for the nested mutex locks case.